src/chainstate/stacks/index/node.rs

```diff
@@ -0,0 +1,2809 @@
+            CursorError::BackptrEncountered(_) => f.write_str(error::Error::description(self)),
+            CursorError::ChrNotFound => f.write_str(error::Error::description(self)),
+        }
+    }
+}
+
+impl error::Error for CursorError {
+    fn cause(&self) -> Option<&error::Error> {
+        None
+    }
+
+    fn description(&self) -> &str {
+        match *self {
+            CursorError::PathDiverged => "Path diverged",
+            CursorError::BackptrEncountered(_) => "Back-pointer encountered",
+            CursorError::ChrNotFound => "Node child not found",
+        }
+    }
+}
+
+/// All numeric values of a Trie node when encoded.
+/// They are all 7-bit numbers -- the 8th bit is used to indicate whether or not the value
+/// identifies a back-pointer to be followed.
+pub mod TrieNodeID {
+    pub const Empty : u8 = 0;
+    pub const Leaf: u8 = 1;
+    pub const Node4 : u8 = 2;
+    pub const Node16 : u8 = 3;
+    pub const Node48 : u8 = 4;
+    pub const Node256 : u8 = 5;
+}
+
+/// A node ID encodes a back-pointer if its high bit is set
+pub fn is_backptr(id: u8) -> bool {
+    id & 0x80 != 0
+}
+
+/// Set the back-pointer bit
+pub fn set_backptr(id: u8) -> u8 {
+    id | 0x80
+}
+
+/// Clear the back-pointer bit
+pub fn clear_backptr(id: u8) -> u8 {
+    id & 0x7f
+}
+
+/// A path in the Trie is the SHA2-512/256 hash of its key.
+pub struct TriePath([u8; 32]);
+impl_array_newtype!(TriePath, u8, 32);
+impl_array_hexstring_fmt!(TriePath);
+impl_byte_array_newtype!(TriePath, u8, 32);
+
+pub const TRIEPATH_MAX_LEN : usize = 32;
+
+impl TriePath {
+    pub fn from_key(k: &String) -> TriePath {
+        let h = TrieHash::from_data(k.as_bytes());
+        let mut hb = [0u8; TRIEPATH_MAX_LEN];
+        hb.copy_from_slice(h.as_bytes());
+        TriePath(hb)
+    }
+}
+
+/// All Trie nodes implement the following methods:
+pub trait TrieNode {
+    /// Node ID for encoding/decoding
+    fn id(&self) -> u8;
+
+    /// Is the node devoid of children?
+    fn empty() -> Self;
+
+    /// Follow a path character to a child pointer
+    fn walk(&self, chr: u8) -> Option<TriePtr>;
+
+    /// Insert a child pointer if the path character slot is not occupied.
+    /// Return true if inserted, false if the slot is already filled
+    fn insert(&mut self, ptr: &TriePtr) -> bool;
+
+    /// Replace an existing child pointer with a new one.  Returns true if replaced; false if the
+    /// child does not exist.
+    fn replace(&mut self, ptr: &TriePtr) -> bool;
+
+    /// Read an encoded instance of this node from a byte stream and instantiate it.
+    fn from_bytes<R: Read>(r: &mut R) -> Result<Self, Error>
+        where Self: std::marker::Sized;
+
+    /// Encode this node instance into a byte stream and write it to the given memory buffer.
+    fn to_bytes(&self, buf: &mut Vec<u8>) -> ();
+
+    /// Encode the consensus-relevant bytes of this node and write it to the given memory buffer.
+    fn to_consensus_bytes(&self, &mut Vec<u8>) -> ();
```

Comment:
`to_consensus_bytes()` returns a serialization of the _consensus-critical_ fields of the node. In practice, this is everything but the `ptr` value of its `TriePtrs`, since the location of the nodes on disk (or in RAM) doesn't need to be agreed upon by nodes in order to correctly handle queries and inserts.  However, the node types, paths, parent/child relationships, and hashes are consensus-critical, since they are what define the structure of the Trie.

That said, `to_consensus_bytes()` only serializes the node's internal data, not its hash or its sibling hashes.  This is important, because the choice of which hash to use is dependent on the MARF structure, not the Trie structure -- i.e. the hash of a child node that lives in a different Trie is the hash of the _block_ in which it resides, as opposed to the hash of the node's data.  This fact is why the hash is _not_ stored within the `TrieNode` implementation.

The fact that a child is in another Trie is captured by `to_consensus_bytes()` already -- the highest bit (`0x80`) of the child's `id` in its `TriePtr` will be set if so, and will be cleared if not.

Hope this helps!